File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack=[]
        for i in s:
            if not stack or i=='(' or i=='{' or i == '[':
                stack.append(i)
            if stack[-1] == '(' and i == ')':
                stack.pop()
            elif stack[-1] == '{' and i == '}':
                stack.pop()
            elif stack[-1] == '[' and i == ']':
                stack.pop()
            elif i == ')' or i == '}' or i == ']':
                stack.append(i)
        if not stack:
            return True
        else:
            return False

        

File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        unordered_map<char,char> mp = {
            {')', '('},
            {'}', '{'},
            {']', '['}
        };

        for (char c : s) {
            if (mp.find(c) != mp.end()) {
                if (st.empty() || st.top() != mp[c])
                    return false;
                st.pop();
            } else {
                st.push(c);
            }
        }

        return st.empty();
    }
};
